### ISSUE TITLE
Redshift migrations break from 0.9.13 to 0.9.14

### ISSUE THREAD

**Issue by jlippold** [db-migrate/node-db-migrate#277]
When running a fresh migration from a blank db on amazon redshift, the script dies

![http://s3.amazonaws.com/PicUp/ed8q35.png](http://s3.amazonaws.com/PicUp/ed8q35.png)

We are currently using 0.9.13 where redshift was still supported. Maybe 0.9.14 added some new postgres command that [redshift doesn't support](http://docs.aws.amazon.com/redshift/latest/dg/c_redshift-and-postgres-sql.html)?

**Comment by wzrdtales** [comment]
@jlippold No there are no new commands in any of the >0.9.13 versions.

Actually the message tells us you're crashing in the all function:
https://github.com/db-migrate/node-db-migrate/blob/9afc4cc480ba762548e5a77f2126cc0bbbceadc7/lib/driver/pg.js#L331

The all function gets only called by System functions, for example when creating a migration table. The query itself doesn't return a result in this case and thus throwing this error.

Can you please rerun this with verbose output and provide everything necessary to reproduce this? Otherwise it's hard to help you.

Thank you

**Comment by wzrdtales** [comment]
@jlippold In v0.9.14 the createMigrationsTable function has been changed. It now uses all and used before runSql, this was not intended and was a bug.

https://github.com/db-migrate/node-db-migrate/blob/918df65a5828b0ee041d388b5088a0458ec5dc15/lib/driver/pg.js#L123-L179

This means one of the following queries was broken even before 0.9.14

```sql
SELECT table_name FROM information_schema.tables WHERE table_name = 'migrations'
SET search_path TO 'public'
SHOW search_path
select version() as version
```

But this would be strange, because if one of these queries fails, then no migration table would be created at all and this also should'nt work with <0.9.13

**Comment by jlippold** [comment]
it's `SET search_path TO $user, public` that's returning the error

```
{ [error: syntax error at or near "$"]
  name: 'error',
  length: 126,
  severity: 'ERROR',
  code: '42601',
  detail: undefined,
  hint: undefined,
  position: '20',
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: '/home/awsrsqa/padb/src/pg/src/backend/parser/parser_scan.l',
  line: '695',
  routine: 'yyerror' }
```
I'm not sure if that helps, i'm digging deeper

**Comment by wzrdtales** [comment]
This error is generated by the lack of quotes. This have to be at least:
SET search_path TO "$user", public

But db-migrate does generate the correct query, I just switched to the 0.9.14 version and tested it. The following query gets generated:
    SET search_path TO test_schema,"$user",public

**Comment by jlippold** [comment]
This sql https://github.com/db-migrate/node-db-migrate/blob/918df65a5828b0ee041d388b5088a0458ec5dc15/lib/driver/pg.js#L148 returns the result differently when using redshift

In the latest postgres `SHOW search_path` returns `"$user",public`
While in redshift `SHOW search_path` returns `$user,public` (no quotes)

Which explains the error.. not too sure the best way to fix just yet

**Comment by wzrdtales** [comment]
ahh, well ok. Now we come close to a solution :)

give me a couple of minutes.

**Comment by jlippold** [comment]
this sql https://github.com/db-migrate/node-db-migrate/blob/918df65a5828b0ee041d388b5088a0458ec5dc15/lib/driver/pg.js#L133 returns this on redshift:
```
PostgreSQL 8.0.2 on i686-pc-linux-gnu, 
compiled by GCC gcc (GCC) 3.4.2 20041017 (Red Hat 3.4.2-6.fc3), 
Redshift 1.0.946
```

and on regular postgres:
```
PostgreSQL 9.3.5 on x86_64-apple-darwin14.0.0, 
compiled by Apple LLVM version 6.0 (clang-600.0.54) (based on LLVM 3.5svn), 
64-bit
```

I suppose if you detect the `Redshift` string in the returned version, you can add the fix based on that condition?

**Comment by wzrdtales** [comment]
https://github.com/db-migrate/node-db-migrate/commit/75e1b17d1468cb21d95887e602d3d757be91eae6

Waiting for travis tests now

**Comment by jlippold** [comment]
that commit works for me on redshift :+1:

**Comment by wzrdtales** [comment]
Going to be published as soon as all travis tests are finished.
Version that is going to hold this fix: 0.9.17

**Comment by jlippold** [comment]
you're the man

**Comment by wzrdtales** [comment]
applied to master, published to v0.9.17
